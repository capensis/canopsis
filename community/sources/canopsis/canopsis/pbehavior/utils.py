"""
Utils for pbehaviors.
"""

from dateutil.rrule import rrulestr
from dateutil import tz
from datetime import datetime


EXDATE_DATE_FORMAT = "%Y/%m/%d %H:%M:%S"


def check_valid_rrule(rrule):
    """
    Check for RRULE validity.

    Raises error when invalid with a more explicit message.
    :param str rrule: rrule as string
    :raises ValueError: rrule is invalid
    :returns: True if ok, raises exception if invalid
    :rtype: bool
    """
    if rrule == '' or rrule is None:
        return True

    try:
        rrulestr(rrule)
    except ValueError as ex:
        raise ValueError('Invalid RRULE: {}'.format(ex))

    return True
