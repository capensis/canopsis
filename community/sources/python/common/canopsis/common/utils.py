# -*- coding: utf-8 -*-
# --------------------------------
# Copyright (c) 2015 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
# ---------------------------------

"""
Python utility library.
"""

from importlib import import_module
from imp import load_source

from inspect import ismodule

from collections import Iterable

from sys import version as PYVER

from os.path import expanduser, splitext
from os.path import join as joinpath
from os import listdir

from re import search as regsearch

from .init import basestring

__RESOLVED_ELEMENTS = {}  #: dictionary of resolved elements by name


#: dictionary which contains singleton per scope
_SINGLETONS_PER_SCOPE = {}


def singleton_per_scope(cls, scope=None, args=None, kwargs=None):
    """Get one instance of ``cls`` per ``scope``.

    :param type cls: class to instanciate.
    :param collections.Hashable scope: key for unique instance of cls.
    :param collections.Iterable args: cls instanciation varargs.
    :param dict kwargs: cls instanciation kwargs.
    :return: singleton of type cls per scope.
    """

    result = None

    # check if an instance has already been created
    if cls in _SINGLETONS_PER_SCOPE and scope in _SINGLETONS_PER_SCOPE[cls]:
        result = _SINGLETONS_PER_SCOPE[cls][scope]

    else:
        # initialiaze both args and kwargs
        if args is None:
            args = ()
        if kwargs is None:
            kwargs = {}
        # instanciate the singleton
        result = cls(*args, **kwargs)
        # register the instance
        _SINGLETONS_PER_SCOPE.setdefault(cls, {})[scope] = result

    return result


def del_singleton_per_scope(cls, scope=None):
    """Delete a singleton of class ``cls`` and ``scope``.

    :param type cls: type of singleton object.
    :param scope: singleton scope.
    """

    _SINGLETONS_PER_SCOPE.get(cls, {}).pop(scope, None)


def dynmodloads(_path='.', subdef=False, pattern='.*', logger=None):
    loaded = {}
    _path = expanduser(_path)

    for mfile in listdir(_path):
        name, ext = splitext(mfile)

        # Ignore "." and "__init__.py" and everything not matched by "*.py"
        if name in ['.', '__init__'] or ext != '.py':
            continue

        logger.info("Load '{0}' ...".format(name))

        try:
            module = load_source(name, joinpath(_path, mfile))

        except ImportError as err:
            logger.error(u'Impossible to import {0}: {1}'.format(name, err))

        else:
            loaded[name] = module

            if subdef:
                alldefs = dir(module)
                builtindefs = [
                    '__builtins__',
                    '__doc__',
                    '__file__',
                    '__name__',
                    '__package__'
                ]

                for mydef in alldefs:
                    if mydef not in builtindefs and regsearch(pattern, mydef):
                        logger.debug(u'from {0} import {1}'.format(
                            name, mydef
                        ))

                        loaded[mydef] = getattr(module, mydef)

    return loaded


def setdefaultattr(obj, attr, value):
    """
    Set attribute in object if not present.

    :param obj: Object to set attribute to
    :type obj: anything

    :param attr: Attribute's name to set
    :type attr: str

    :param value: Value to set
    :type value: anything

    :returns: current value if attribute exists, or new value otherwise
    """

    if hasattr(obj, attr):
        return getattr(obj, attr)

    else:
        setattr(obj, attr, value)
        return value


def free_cache(path=None):
    """
    Remove an element from cache memory

    :param str path: path to element to remove from cache. If None, remove all
        elements from cache.
    """

    if path is None:
        __RESOLVED_ELEMENTS = {}
    else:
        __RESOLVED_ELEMENTS[path].pop(path, '')


def lookup(path, cached=True):
    """
    Get element reference from input full path element.

    :limitations: does not resolve class method.

    :param str path: full path to a python element.
        Examples:

        - __builtin__.open
        - canopsis.common.utils.lookup

    :para bool cached: if True (by default), use __RESOLVED_ELEMENTS cache
        memory to quickly load elements

    :return: python object which is accessible thourgh input path.
    :rtype: object
    """

    path = expanduser(path)
    element_in_cache = cached and path in __RESOLVED_ELEMENTS

    result = __RESOLVED_ELEMENTS[path] if element_in_cache else None

    if result is None and path:

        components = path.split('.')
        index = 0
        components_len = len(components)

        module_name = components[0]

        # try to import the first component name
        try:
            result = import_module(module_name)
        except ImportError as e:
            print(
                'Error while importing module {} : {}'.format(module_name, e)
            )

        if result is not None:

            if components_len > 1:

                index = 1

                # try to import all sub-modules/packages
                try:  # check if name is defined from an external module
                    # find the right module

                    while index < components_len:
                        module_name = '{0}.{1}'.format(
                            module_name, components[index])
                        result = import_module(module_name)
                        index += 1

                except ImportError as ie:
                    # path sub-module content
                    try:

                        while index < components_len:
                            result = getattr(result, components[index])
                            index += 1

                    except AttributeError as ae:
                        raise ImportError(
                            ('Wrong path {0} at {1}, ' +
                                'errors when importing module {2} ' +
                                ': {3}, {4}').format(
                                path,
                                components[:index],
                                module_name,
                                ie,
                                ae
                            ))

            if result is not None and cached:
                __RESOLVED_ELEMENTS[path] = result

        else:  # get relative object from current module
            raise ImportError(
                'Does not handle relative path: {0}'.format(path)
            )

    return result


def path(element):
    """
    Get full path of a given element.

    Do the inverse of lookup

    :param element: must be directly defined into a module or a package
    :type element: object
    """

    if not hasattr(element, '__name__'):
        raise AttributeError(
            'element {0} must have the attribute __name__'.format(element))

    result = element.__name__ if ismodule(element) else \
        '{0}.{1}'.format(element.__module__, element.__name__)

    return result


def isiterable(element, is_str=True):
    """
    Check whatever or not if input element is an iterable.

    :param is_str: check if element is also a str
    :type is_str: bool
    """
    result = isinstance(element, Iterable) \
        and (is_str or not isinstance(element, basestring))

    return result


def isunicode(s):
    """
    Check if string is unicode.

    :param s: string to check
    :type s: basestring

    :return: True if unicode or Python3, False otherwise
    """

    if PYVER < '3':
        return isinstance(s, unicode)

    else:
        return True


def ensure_unicode(s):
    """Convert string to unicode.

    :param s: string to convert
    :type s: basestring

    :return: unicode (or the same string if Python3)
    """

    result = s

    if PYVER < '3':
        if isinstance(s, basestring):
            if not isinstance(s, unicode):
                result = s.decode()
        else:
            raise TypeError('Expecting a string as argument')

    return result


def forceUTF8(data, _memory=None):
    """Return a copy of data where all embedded strings are UTF8.

    :param data: data from where convert str to UTF8 format.
    :return: data copy where all str are in UTF8 format.
    """

    # by default, result is data
    result = data
    # do something only if python version is 2
    if PYVER < '3':
        # initialize memory
        if _memory is None:
            _memory = {}
        # if data has already been processed
        data_id = id(data)
        if data_id in _memory:
            # result is the previous result
            result = _memory[data_id]
        else:  # else process data
            # if data is a basestring, decode it to an utf8
            if isinstance(data, basestring):
                if not isinstance(data, unicode):
                    result = data.decode('utf-8', 'ignore')
            # if data is a dict
            elif isinstance(data, dict):
                data_id = id(data)
                if data_id in _memory:
                    result = _memory[data_id]
                else:
                    # copy data
                    result = data.copy()
                    for param in data:
                        value = data[param]
                        # convert param and value
                        param = forceUTF8(param)
                        value = forceUTF8(value)
                        result[param] = value
            # if data is an iterable
            elif isinstance(data, Iterable):
                result = []
                # convert all values of data
                for d in data:
                    value = forceUTF8(d)
                    result.append(value)
                # and convert result to data type
                result = type(data)(result)
            # save the result in memory
            _memory[data_id] = result

    return result


def ensure_iterable(value, iterable=list):
    """
    Convert a value into an iterable if it is not.

    :param value: value to convert
    :type value: object

    :param iterable: iterable type to apply (default: list)
    :type iterable: type
    """

    result = value

    if not isiterable(value, is_str=False) or isinstance(value, dict):
        result = [value]
        result = iterable(result)

    else:
        result = iterable(value)

    return result


def get_first(iterable, default=None):
    """
    Try to get input iterable first item or default if iterable is empty.
    """

    result = iterable[0] if iterable else default

    return result


def prototype(typed_args=None, typed_kwargs=None, typed_return=None):
    """
    Decorate a function to check its parameters type.

    :param typed_args: Types for args
    :type typed_args: tuple

    :param typed_kwargs: Types for kwargs
    :type typed_kwargs: dict

    :param typed_return: Types for return
    :type typed_return: tuple of type, or type

    :raises: TypeError
    """

    if typed_args is None:
        typed_args = ()

    if typed_kwargs is None:
        typed_kwargs = {}

    if typed_return is None:
        typed_return = type(None)

    if isinstance(typed_return, list):
        typed_return = tuple(typed_return)

    def decorator(func):
        def wrapper(*args, **kwargs):
            i = 0
            l = len(args)

            for i in range(l):
                types = typed_args[i]

                if isinstance(types, list):
                    types = tuple(types)

                if not isinstance(args[i], types):
                    raise TypeError(
                        'Invalid type for arg#{}, got {} instead of {}'.format(
                            type(args[i]),
                            types[i]
                        )
                    )

            for key in typed_kwargs:
                types = typed_kwargs[key]

                if isinstance(types, list):
                    types = tuple(types)

                arg = kwargs.get(key, None)

                if not isinstance(arg, types):
                    raise TypeError(
                        'Invalid type for {}, got {} instead of {}'.format(
                            key,
                            type(arg),
                            types
                        )
                    )

            ret = func(*args, **kwargs)

            if not isinstance(ret, typed_return):
                raise TypeError(
                    'Invalid type for return, got {0} instead of {}'.format(
                        type(ret),
                        typed_return
                    )
                )

            return ret

        return wrapper

    return decorator


class dictproperty(object):
    """
        Property decorator for dict-like attributes.
    """

    class _proxy(object):
        """
            Proxy interface to dict-like objects.
        """

        def __init__(self, obj, fget, fset, fdel, *args, **kwargs):
            super(dictproperty._proxy, self).__init__(*args, **kwargs)

            self._obj = obj
            self._fget = fget
            self._fset = fset
            self._fdel = fdel

        def __getitem__(self, key):
            if self._fget is None:
                raise TypeError("Impossible to get key: {0}".format(key))

            return self._fget(self._obj, key)

        def __setitem__(self, key, value):
            if self._fset is None:
                raise TypeError("Impossible to set key: {0} = {1}".format(key, value))

            self._fset(self._obj, key, value)

        def __delitem__(self, key):
            if self._fdel is None:
                raise TypeError("Impossible to delete key: {0}".format(key))

            self._fdel(self._obj, key)

    def __init__(self, fget=None, fset=None, fdel=None, doc=None, *args, **kwargs):
        super(dictproperty, self).__init__(*args, **kwargs)

        self._fget = fget
        self._fset = fset
        self._fdel = fdel
        self.__doc__ = doc

    def __get__(self, obj, objtype=None):
        if obj is None:
            return self

        return self._proxy(obj, self._fget, self._fset, self._fdel)
