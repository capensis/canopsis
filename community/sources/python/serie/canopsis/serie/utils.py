# -*- coding: utf-8 -*-
# --------------------------------
# Copyright (c) 2015 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
# ---------------------------------

"""Module of Serie utilities."""

NAME_BY_KEYS = {
    'co:': 'component',
    're:': 'resource',
    'me:': 'name'
}  #: entity name by regex entity header.


def build_filter_from_regex(regex):
    """
    Transform a metric filter into a MongoDB-like filter.

    co:<regex> --> {'component': {'$regex': '<regex>'}}
    re:<regex> --> {'resource': {'$regex': '<regex>'}}
    me:<regex> --> {'metric': {'$regex': '<regex>'}}

    co:<regex1> co:<regex2> --> {'$or': [...]}
    co:<regex> re:<regex> --> {'$and': [...]}

    :param regex: Metric filter to transform
    :type regex: str

    :returns: MongoDB-like filter as dict
    """

    regex_parts = regex.split(' ')

    regex = {NAME_BY_KEYS[key]: [] for key in NAME_BY_KEYS}

    for part in regex_parts:
        for key in NAME_BY_KEYS:
            if part.startswith(key):
                name = NAME_BY_KEYS[key]
                regex[name].append({'$regex': part[len(key):]})
                break

        else:
            for name in regex.keys():
                regex[name].append({'$regex': part})

    mfilter = {'$and': []}

    for key in regex:
        items = regex[key]
        if items:
            local_mfilter = {'$or': [
                {key: subfilter} for subfilter in items
            ]}

            if len(local_mfilter['$or']) == 1:
                local_mfilter = local_mfilter['$or'][0]

            mfilter['$and'].append(local_mfilter)

    if len(mfilter['$and']) == 0:
        mfilter = {}

    elif len(mfilter['$and']) == 1:
        mfilter = mfilter['$and'][0]

    return mfilter
