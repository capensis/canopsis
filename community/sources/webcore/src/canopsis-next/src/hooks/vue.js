import { getCurrentInstance } from 'vue';

/**
 * Hook for get current component instance
 *
 * @return {Vue}
 */
export const useComponentInstance = () => {
  const instance = getCurrentInstance();

  return instance?.proxy || instance;
};

/**
 * Hook for get current component options
 *
 * @return {Object}
 */
export const useComponentOptions = () => {
  const instance = useComponentInstance();

  return instance.$options;
};

/**
 * Hook for get current component model
 *
 * @return {Object}
 */
export const useComponentModel = () => {
  const { model } = useComponentOptions();

  return {
    event: model?.event ?? 'input',
    prop: model?.event ?? 'value',
  };
};
