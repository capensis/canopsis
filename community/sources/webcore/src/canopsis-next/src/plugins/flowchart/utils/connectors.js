import { range } from 'lodash';

import uid from '@/helpers/uid';

/**
 * @typedef {Object} ConnectorPoint
 * @property {number} x
 * @property {number} y
 * @property {string} _id
 */

/**
 * Generate connector
 *
 * @param {number} x
 * @param {number} y
 * @returns {ConnectorPoint}
 */
export const generateConnectorPoint = (x, y) => ({
  x,
  y,
  _id: uid(),
});

/**
 * Generate connector points by rectangle
 *
 * @param {Rect} rect
 * @param {number} count
 * @returns {ConnectorPoint[]}
 */
export const generateConnectorsByRect = (rect, count) => {
  const { width, height, rx, ry, x, y } = rect;
  const sideConnectorsCount = Math.round(count / 4);
  const sideSpacesCount = sideConnectorsCount + 1;
  const startX = x + rx;
  const startY = x + ry;

  const widthConnectorOffset = (width - rx * 2) / (sideSpacesCount);
  const heightConnectorOffset = (height - ry * 2) / (sideSpacesCount);

  const sideConnectors = range(1, sideConnectorsCount + 1);

  const topConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y,
  ));

  const rightConnectors = sideConnectors.map(index => generateConnectorPoint(
    x + width,
    startY + heightConnectorOffset * index,
  ));

  const bottomConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y + height,
  ));

  const leftConnectors = sideConnectors.map(index => generateConnectorPoint(
    x,
    startY + heightConnectorOffset * index,
  ));

  const connectors = [
    ...topConnectors,
    ...rightConnectors,
    ...bottomConnectors,
    ...leftConnectors,
  ];

  if (!rx && !ry) {
    /* Angles connectors */
    connectors.push(
      generateConnectorPoint(x, y),
      generateConnectorPoint(x + width, y),
      generateConnectorPoint(x + width, y + height),
      generateConnectorPoint(x, y + height),
    );
  }

  return connectors;
};
