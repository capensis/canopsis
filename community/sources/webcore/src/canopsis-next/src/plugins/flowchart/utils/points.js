/**
 * Calculate points with ghosts points
 *
 * @param {Array} points
 * @returns {Array}
 */
export const getPointsWithGhosts = points => points.reduce((acc, point, index) => {
  const nextIndex = index + 1;
  const nextPoint = points[nextIndex];

  acc.push({ ...point, index });

  if (nextPoint) {
    acc.push({
      x: (point.x + nextPoint.x) / 2,
      y: (point.y + nextPoint.y) / 2,
      ghost: true,
      index: nextIndex,
    });
  }

  return acc;
}, []);
