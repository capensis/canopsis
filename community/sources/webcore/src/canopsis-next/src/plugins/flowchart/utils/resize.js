import { DIRECTIONS } from '@/plugins/flowchart/constants';

/**
 * @typedef {Object} Rect
 * @property {number} width
 * @property {number} height
 * @property {number} x
 * @property {number} y
 */

/**
 * @typedef {Object} Square
 * @property {number} size
 * @property {number} x
 * @property {number} y
 */

/**
 * @typedef {Object} Cursor
 * @property {number} x
 * @property {number} y
 */

/**
 * Resize rectangle by direction and cursor position
 *
 * @param {Rect} rect
 * @param {Cursor} cursor
 * @param {string} direction
 * @returns {{ rect: Rect, direction: string }}
 */
export const resizeRectangleShapeByDirection = (rect, cursor, direction) => {
  const { x: cursorX, y: cursorY } = cursor;
  const newRect = {
    x: rect.x,
    y: rect.y,
    width: rect.width,
    height: rect.height,
  };
  const directionArray = direction.split('');

  directionArray.forEach((singleDirection, index) => {
    switch (singleDirection) {
      case DIRECTIONS.south: {
        const newHeight = cursorY - newRect.y;

        if (newHeight > 0) {
          newRect.height = newHeight;
        } else {
          newRect.height = Math.abs(newHeight);
          newRect.y -= newRect.height;

          directionArray[index] = DIRECTIONS.north;
        }

        break;
      }
      case DIRECTIONS.north: {
        const newHeight = newRect.height + newRect.y - cursorY;

        if (newHeight > 0) {
          newRect.height = newHeight;
          newRect.y = cursorY;
        } else {
          newRect.height = Math.abs(newHeight);
          newRect.y = cursorY - newRect.height;

          directionArray[index] = DIRECTIONS.south;
        }

        break;
      }
      case DIRECTIONS.east: {
        const newWidth = cursorX - newRect.x;

        if (newWidth > 0) {
          newRect.width = newWidth;
        } else {
          newRect.width = Math.abs(newWidth);
          newRect.x = cursorX;

          directionArray[index] = DIRECTIONS.west;
        }

        break;
      }
      case DIRECTIONS.west: {
        const newWidth = newRect.width + newRect.x - cursorX;

        if (newWidth > 0) {
          newRect.width = newWidth;
          newRect.x = cursorX;
        } else {
          newRect.width = Math.abs(newWidth);
          newRect.x = cursorX - newRect.width;

          directionArray[index] = DIRECTIONS.east;
        }

        break;
      }
    }
  });

  return {
    rect: newRect,
    direction: directionArray.join(''),
  };
};

/**
 * Resize square by direction and cursor position
 *
 * @param {Square} square
 * @param {Cursor} cursor
 * @param {string} direction
 * @returns {{ square: Square, direction: string }}
 */
export const resizeSquareShapeByDirection = (square, cursor, direction) => {
  const { x: cursorX, y: cursorY } = cursor;
  const newSquare = { ...square };
  let newDirection = direction;

  switch (direction) {
    case DIRECTIONS.west:
    case DIRECTIONS.southWest: {
      const newSize = square.size + square.x - cursorX;

      if (newSize >= 0) {
        newSquare.size = newSize;
        newSquare.x = cursorX;
      } else {
        const absoluteNewSize = Math.abs(newSize);

        newSquare.x += square.size;
        newSquare.y -= absoluteNewSize;
        newSquare.size = absoluteNewSize;

        newDirection = DIRECTIONS.northEast;
      }

      break;
    }
    case DIRECTIONS.east:
    case DIRECTIONS.northEast: {
      const newSize = cursorX - square.x;

      if (newSize >= 0) {
        newSquare.y += square.size - newSize;
        newSquare.size = newSize;
      } else {
        const absoluteNewSize = Math.abs(newSize);

        newSquare.y += square.size;
        newSquare.x -= absoluteNewSize;
        newSquare.size = absoluteNewSize;

        newDirection = DIRECTIONS.southWest;
      }

      break;
    }
    case DIRECTIONS.south:
    case DIRECTIONS.southEast: {
      const newSize = cursorY - square.y;

      if (newSize >= 0) {
        newSquare.size = newSize;
      } else {
        const absoluteNewSize = Math.abs(newSize);

        newSquare.size = absoluteNewSize;
        newSquare.x -= absoluteNewSize;
        newSquare.y -= absoluteNewSize;

        newDirection = DIRECTIONS.northWest;
      }

      break;
    }
    case DIRECTIONS.north:
    case DIRECTIONS.northWest: {
      const newSize = square.y + square.size - cursorY;

      if (newSize >= 0) {
        const diffBetweenSizes = square.size - newSize;

        newSquare.size = newSize;
        newSquare.y += diffBetweenSizes;
        newSquare.x += diffBetweenSizes;
      } else {
        const absoluteNewSize = Math.abs(newSize);

        newSquare.y += square.size;
        newSquare.x += square.size;
        newSquare.size = absoluteNewSize;

        newDirection = DIRECTIONS.southEast;
      }

      break;
    }
  }

  return {
    square: newSquare,
    direction: newDirection,
  };
};
