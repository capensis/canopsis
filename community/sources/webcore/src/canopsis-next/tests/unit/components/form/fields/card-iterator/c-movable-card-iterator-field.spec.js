import { generateRenderer, generateShallowRenderer } from '@unit/utils/vue';

import { createButtonStub } from '@unit/stubs/button';
import CMovableCardIteratorField from '@/components/forms/fields/card-iterator/c-movable-card-iterator-field.vue';
import MovableCardIteratorItem from '@/components/forms/fields/card-iterator/movable-card-iterator-item.vue';

const stubs = {
  'movable-card-iterator-item': MovableCardIteratorItem,
  'v-btn': createButtonStub('v-btn'),
};

const selectCards = wrapper => wrapper.findAll('v-card-stub');
const selectCardByIndex = (wrapper, index) => selectCards(wrapper).at(index);
const selectCardButtons = wrapper => wrapper.findAll('.v-btn');
const selectCardUpButton = wrapper => selectCardButtons(wrapper).at(0);
const selectCardDownButton = wrapper => selectCardButtons(wrapper).at(1);
const selectCardRemoveButton = wrapper => selectCardButtons(wrapper).at(2);

describe('c-movable-card-iterator-field', () => {
  const items = [
    { key: 'key-1', title: 'title-1' },
    { key: 'key-2', title: 'title-2' },
    { key: 'key-3', title: 'title-3' },
    { key: 'key-4', title: 'title-4' },
  ];

  const factory = generateShallowRenderer(CMovableCardIteratorField, { stubs });
  const snapshotFactory = generateRenderer(CMovableCardIteratorField);

  test('Card moved below after click on down button', async () => {
    const wrapper = factory({
      propsData: {
        items,
      },
    });

    const movingCard = selectCardByIndex(wrapper, 1);

    selectCardDownButton(movingCard).trigger('click');

    expect(wrapper).toEmit('input', [
      items[0],
      items[2],
      items[1],
      items[3],
    ]);
  });

  test('Card doesn\'t moved below after try to move down last card', async () => {
    const wrapper = factory({
      propsData: {
        items,
      },
    });

    const movingCard = selectCardByIndex(wrapper, 3);

    selectCardDownButton(movingCard).trigger('click');

    expect(wrapper).not.toEmit('input');
  });

  test('Card moved above after click on up button', async () => {
    const wrapper = factory({
      propsData: {
        items,
      },
    });

    const movingCard = selectCardByIndex(wrapper, 3);

    selectCardUpButton(movingCard).trigger('click');

    expect(wrapper).toEmit('input', [
      items[0],
      items[1],
      items[3],
      items[2],
    ]);
  });

  test('Card doesn\'t moved above after try to move up first card', async () => {
    const wrapper = factory({
      propsData: {
        items,
      },
    });

    const movingCard = selectCardByIndex(wrapper, 0);

    selectCardUpButton(movingCard).trigger('click');

    expect(wrapper).not.toEmit('input');
  });

  test('Card removed above after click on remove button', async () => {
    const wrapper = factory({
      propsData: {
        items,
      },
    });

    const removingCard = selectCardByIndex(wrapper, 2);

    selectCardRemoveButton(removingCard).trigger('click');

    expect(wrapper).toEmit('input', [
      items[0],
      items[1],
      items[3],
    ]);
  });

  test('Renders `c-movable-card-iterator-field` with default props', async () => {
    const wrapper = snapshotFactory();

    expect(wrapper).toMatchSnapshot();
  });

  test('Renders `c-movable-card-iterator-field` with custom props', async () => {
    const wrapper = snapshotFactory({
      propsData: {
        items: [
          { id: 'id-1', text: 'text-1' },
          { id: 'id-2', text: 'text-2' },
        ],
        itemKey: 'id',
        addable: true,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  test('Renders `c-movable-card-iterator-field` with slots props', async () => {
    const wrapper = snapshotFactory({
      propsData: {
        items: [
          { id: 'id-1', text: 'text-1' },
          { id: 'id-2', text: 'text-2' },
        ],
        itemKey: 'id',
        addable: true,
      },
      slots: {
        prepend: '<div class="prepend-slot" />',
        append: '<div class="append-slot" />',
      },
      scopedSlots: {
        item(props) {
          return this.$createElement(
            'div',
            { attrs: { class: 'item-slot' } },
            JSON.stringify(props),
          );
        },
      },
    });

    expect(wrapper).toMatchSnapshot();
  });
});
