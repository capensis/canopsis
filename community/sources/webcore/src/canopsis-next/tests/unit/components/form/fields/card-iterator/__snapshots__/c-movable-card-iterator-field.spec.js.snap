// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-movable-card-iterator-field Renders \`c-movable-card-iterator-field\` with custom props 1`] = `
<div>
  <div
    class="my-2 v-card v-sheet theme--light"
  >
    <div
      class="layout pt-2 justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <span
          class="pl-2"
        >
          <button
            class="v-btn v-btn--disabled v-btn--icon theme--light"
            disabled="disabled"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_upward
              </i>
            </div>
          </button>
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_downward
              </i>
            </div>
          </button>
        </span>
      </div>
      <div
        class="flex text-xs-right pr-2 xs3"
      >
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light error--text"
            >
              close
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      class="layout px-3 pb-3 justify-center column"
    />
  </div>
  <div
    class="my-2 v-card v-sheet theme--light"
  >
    <div
      class="layout pt-2 justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <span
          class="pl-2"
        >
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_upward
              </i>
            </div>
          </button>
          <button
            class="v-btn v-btn--disabled v-btn--icon theme--light"
            disabled="disabled"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_downward
              </i>
            </div>
          </button>
        </span>
      </div>
      <div
        class="flex text-xs-right pr-2 xs3"
      >
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light error--text"
            >
              close
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      class="layout px-3 pb-3 justify-center column"
    />
  </div>
  <div
    class="layout row wrap"
  >
    <button
      class="mx-0 v-btn theme--light primary"
      type="button"
    >
      <div
        class="v-btn__content"
      >
        Add
      </div>
    </button>
  </div>
</div>
`;

exports[`c-movable-card-iterator-field Renders \`c-movable-card-iterator-field\` with default props 1`] = `
<div>
  <div
    class="layout row wrap"
  >
    <button
      class="mx-0 v-btn theme--light primary"
      type="button"
    >
      <div
        class="v-btn__content"
      >
        Add
      </div>
    </button>
  </div>
</div>
`;

exports[`c-movable-card-iterator-field Renders \`c-movable-card-iterator-field\` with slots props 1`] = `
<div>
  <div
    class="prepend-slot"
  />
  <div
    class="my-2 v-card v-sheet theme--light"
  >
    <div
      class="layout pt-2 justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <span
          class="pl-2"
        >
          <button
            class="v-btn v-btn--disabled v-btn--icon theme--light"
            disabled="disabled"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_upward
              </i>
            </div>
          </button>
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_downward
              </i>
            </div>
          </button>
        </span>
      </div>
      <div
        class="flex text-xs-right pr-2 xs3"
      >
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light error--text"
            >
              close
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      class="layout px-3 pb-3 justify-center column"
    >
      <div
        class="item-slot"
      >
        {"item":{"id":"id-1","text":"text-1"},"index":0}
      </div>
    </div>
  </div>
  <div
    class="my-2 v-card v-sheet theme--light"
  >
    <div
      class="layout pt-2 justify-space-between"
    >
      <div
        class="flex xs5"
      >
        <span
          class="pl-2"
        >
          <button
            class="v-btn v-btn--icon theme--light"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_upward
              </i>
            </div>
          </button>
          <button
            class="v-btn v-btn--disabled v-btn--icon theme--light"
            disabled="disabled"
            type="button"
          >
            <div
              class="v-btn__content"
            >
              <i
                aria-hidden="true"
                class="v-icon material-icons theme--light"
              >
                arrow_downward
              </i>
            </div>
          </button>
        </span>
      </div>
      <div
        class="flex text-xs-right pr-2 xs3"
      >
        <button
          class="v-btn v-btn--icon theme--light"
          type="button"
        >
          <div
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon material-icons theme--light error--text"
            >
              close
            </i>
          </div>
        </button>
      </div>
    </div>
    <div
      class="layout px-3 pb-3 justify-center column"
    >
      <div
        class="item-slot"
      >
        {"item":{"id":"id-2","text":"text-2"},"index":1}
      </div>
    </div>
  </div>
  <div
    class="append-slot"
  />
  <div
    class="layout row wrap"
  >
    <button
      class="mx-0 v-btn theme--light primary"
      type="button"
    >
      <div
        class="v-btn__content"
      >
        Add
      </div>
    </button>
  </div>
</div>
`;
